File: leetcode/easy/dp/70_climbing_stairs.py
from collections import defaultdict

class Solution(object):
    # brute force
    # TLE
    def climbStairsBrute(self, n):
        """
        :type n: int
        :rtype: int
        """
        # create a tree and then traverse it to so how many paths lead to n
        tree = defaultdict(list)
        count = [0]
        for i in range(n):
            tree[i].append(i+1)
            tree[i].append(i+2)
        print(tree)
        self.dfs(0, tree, count, n)
        return count[0]
    def dfs(self, node, tree, count, n):
        if node==n:
            count[0] += 1
            return
        for i in tree[node]:
            self.dfs(i, tree, count, n)

    # DP
    # we are solving similar problems again and again, see the print of above tree and draw it
    # if we already have 2 we don't need to do a complete DFS and calculation, similarly for 3 as well
    # we first start with finding the bottom-most element and work our way up
    # this is called bottom up DP
    # therefore to reach n from DP[n]  there is only 1 way, thus DP[n] = 1 also from n-1 only possible way to reach n is 1
    # thus DP[n-1] = 1
    def climbStairs(self, n):
        dp = [1]*(n+1)
        for i in reversed(range(0, n-1)):
            dp[i] = dp[i+1]+dp[i+2]
        return max(dp)
    
    # optimized space
    # we are basically adding the last two numbers to itself
    # thus switching the second number to newly added one would make sense
    # somehow this says it only beats 10% while the one above beats 90, weird.
    def climbingStairsOptimized(self, n):
        first, second = 1, 1
        for i in range(0, n-1):
            temp = second
            second = first+second
            first = temp
        return second
    
if __name__=='__main__':
    n = 5
    s = Solution()
    # print(s.climbStairsBrute(n))
    print(s.climbingStairsOptimized(n))

File: NeetCode RoadMap/DP_1D/easy/70_climbing_stairs.cpp
#include<iostream>
#include<unordered_map>
#include<vector>

using namespace std;

class Solution {
public:
    // literally fibo
    int climbStairs(int n) {
        if(n==1) {
            return 1;
        }
        if(n==2) {
            return 2;
        }
        int one = 1;
        int two = 2;
        int three = 0;
        for (int i = 1; i < n-1; i++)
        {
            three = one+two;
            one = two;
            two = three;
        }
        return three;
    }

public:
    // recursive fibo, TLE
    int climbStairsRecurs(int n) {
        if(n==1) {
            return 1;
        }
        if(n==2) {
            return 2;
        }
        return climbStairsRecurs(n-1) + climbStairsRecurs(n-2);
    }

/*
    create a tree, start with step=0 then there are two possibilities, take 2 steps or take 1 step, and then 2 possibilities for each child
*/
public:
    int climbStairsTree(int n) {
        if (n==1)
        {
            return 1;
        }
        if (n==2)
        {
            return 2;
        }
        unordered_map<int, vector<int>> tree;
        vector<int> count = {0};

        for (int i = 0; i < n; i++)
        {
            tree[i] = {i+1, i+2};
        }
        dfs(count, 0, n, tree);

        return count[0];
    }
public:
    void dfs(vector<int>& count, int node, int n, unordered_map<int, vector<int>>& tree) {
        if(node == n){
            count[0]++;
            return;
        }

        vector<int> temp = tree[node];
        for(int currNode: temp) {
            dfs(count, currNode, n, tree);
        }
    }

/*DP solution, basically fibo but with an array*/
public:
    int climbStairsDP(int n) {
        if (n==1)
        {
            return 1;
        }
        if (n==2)
        {
            return 2;
        }
        vector<int> dp(n+1, 1);
        for(int i = n-2; i >= 0; i--) {
            dp[i] = dp[i+1] + dp[i+2];
        }
        for(int it : dp) {
            cout << it << " ";
        }
        cout << "\n";
        return dp[0];
    }
};

int main() {
    Solution sol;
    cout << sol.climbStairsDP(5);
    return 0;
}